### DISCUSSION TITLE
is there a way to apply middleware to whole router?

### DISCUSSION THREAD

**Discussion by capaj** [trpc/trpc#3579] [Q&A] [answered]
per title

**Comment by KATT** [comment, marked as answer]
**tl;dr:** only by using base procedures: https://trpc.io/docs/procedures#reusable-base-procedures

What I experienced with v9 - where we did it at a route level - is that I often wanted to opt out of middleware(s) deep in a router somewhere and that structure makes it next to impossible.

I want to have "domain logic" in the same space: if I have a "post" router there might be some routes that are public like (read) and some that are protected like create/update.
